### PR TITLE
Hw

### PR DESCRIPTION
**Pull request by Dinexpod**

### CODE REVIEW

File: src/main/java/mate/academy/jpahw/devices/Device.java [hunk 1]
```diff
@@ -0,0 +1,103 @@
...
+    private Integer id;
+    @Column
+    private String name;
+    @Column
+    private String model;
+    @Column
+    private String brand;
+    boolean avaliableAcsessory = false;
+
+    protected Device turnOnAcsessory(Object acsessory) {
```
Comment (boroda4436) [inline src/main/java/mate/academy/jpahw/devices/Device.java]: Why you are passing acsessory object? It is not used in this method

File: src/main/java/mate/academy/jpahw/devices/Device.java [hunk 2]
```diff
@@ -0,0 +1,103 @@
...
+        this.avaliableAcsessory = true;
+        return this;
+    }
+
+    protected Device turnOffAcsessory(Object acsessory) {
+        this.avaliableAcsessory = false;
+        return this;
+    }
+
+    public Test doTest(Patient patient) {
```
Comment (boroda4436) [inline src/main/java/mate/academy/jpahw/devices/Device.java]: let's move this to service

File: src/main/java/mate/academy/jpahw/devices/Photometer.java [hunk 3]
```diff
@@ -0,0 +1,135 @@
...
+    @Column(name = "qr_code")
+    private String qrCode;
+    @Column
+    private Double cost;
+    @Column
+    @OneToMany(mappedBy = "ownerPhotometer", fetch = FetchType.EAGER)
+    protected List<PhotometerAcsessory> photometerAcsessories = new ArrayList<>(1);
+
+    @Override
+    public Test doTest(Patient patient) {
```
Comment (boroda4436) [inline src/main/java/mate/academy/jpahw/devices/Photometer.java]: let's move this to service

File: src/main/java/mate/academy/jpahw/devices/Photometer.java [hunk 4]
```diff
@@ -0,0 +1,135 @@
...
+        System.out.println("Photometer acsessory turned ON");
+        return this;
+    }
+
+    @Override
+    public Device turnOffAcsessory(Object acsessory) {
+        PhotometerAcsessory photometerAcsessory = (PhotometerAcsessory) acsessory;
+        photometerAcsessories.remove(photometerAcsessory);
+        photometerAcsessory.setState(UNAPPLIED);
+        System.out.println("Photometer acsessory turned OFF");
```
Comment (boroda4436) [inline src/main/java/mate/academy/jpahw/devices/Photometer.java]: let's log information after method invokation, not inside method

File: src/main/java/mate/academy/jpahw/devices/UltrasonicDevice.java [hunk 5]
```diff
@@ -0,0 +1,137 @@
...
+    private Integer id;
+    @Column
+    private Integer weight;
+    @Column(name = "qr_code")
+    private String qrCode;
+    @Column
+    private Double cost;
+    @Column
+    @OneToMany(fetch = FetchType.LAZY, mappedBy = "ownerUltrasonic")
+    private List<UltrasonicAcsessory> ultrasonicAcsessories = new ArrayList<>(1);
```
Comment (boroda4436) [inline src/main/java/mate/academy/jpahw/devices/UltrasonicDevice.java]: Let me think if it's better to use Accessoty list here or not. Ping me offline with this question

File: src/main/java/mate/academy/jpahw/patients/Patient.java [hunk 6]
```diff
@@ -0,0 +1,86 @@
...
+import javax.persistence.GenerationType;
+import javax.persistence.Id;
+import javax.persistence.Inheritance;
+import javax.persistence.InheritanceType;
+import java.util.Date;
+
+@Entity(name = "patient")
+@Inheritance(strategy = InheritanceType.JOINED)
+public class Patient {
+    private static EntityManager em = Main.getEm();
```
Comment (boroda4436) [inline src/main/java/mate/academy/jpahw/patients/Patient.java]: Why you need EntityManager in entity? It's hardcoding

File: src/main/java/mate/academy/jpahw/patients/Patient.java [hunk 7]
```diff
@@ -0,0 +1,86 @@
...
+
+    public Patient() {
+    }
+
+    public Patient(String name, String lastName, EntityManager em) {
+        this.name = name;
+        this.lastName = lastName;
+        date = new Date();
+
+        em.getTransaction().begin();
```
Comment (boroda4436) [inline src/main/java/mate/academy/jpahw/patients/Patient.java]: storing in DB should be in DAO layer

File: src/main/java/mate/academy/jpahw/patients/PatientService.java [hunk 8]
```diff
@@ -0,0 +1,29 @@
...
+    public List<Test> getAllTests() {
+        TypedQuery<Test> query = em
+                .createQuery("select p from tests p ", Test.class);
+        return query.getResultList();
+    }
+
+    public List<Test> getAllTestsInDateRange(LocalDateTime from, LocalDateTime to) {
+        TypedQuery<Test> query = em
+                .createQuery("select p from tests p ", Test.class);
+        return query.getResultList()
```
Comment (boroda4436) [inline src/main/java/mate/academy/jpahw/patients/PatientService.java]: +1

File: src/main/java/mate/academy/jpahw/devices/Device.java (hunk 2 above)
Comment (Dinexpod) [reply src/main/java/mate/academy/jpahw/devices/Device.java]: ok

File: src/main/java/mate/academy/jpahw/devices/Photometer.java (hunk 3 above)
Comment (Dinexpod) [reply src/main/java/mate/academy/jpahw/devices/Photometer.java]: ok

File: src/main/java/mate/academy/jpahw/devices/Photometer.java (hunk 4 above)
Comment (Dinexpod) [reply src/main/java/mate/academy/jpahw/devices/Photometer.java]: ok

File: src/main/java/mate/academy/jpahw/patients/Patient.java (hunk 7 above)
Comment (Dinexpod) [reply src/main/java/mate/academy/jpahw/patients/Patient.java]: Ok

File: src/main/java/mate/academy/jpahw/EntityManagerHolder.java [hunk 9]
```diff
@@ -0,0 +1,19 @@
+package mate.academy.jpahw;
+
+import javax.persistence.EntityManager;
+import javax.persistence.EntityManagerFactory;
+import javax.persistence.Persistence;
+
+public class EntityManagerHolder {
```
Comment (boroda4436) [inline src/main/java/mate/academy/jpahw/EntityManagerHolder.java]: https://hackernoon.com/singleton-pattern-but-dont-get-too-comfortable-3ae69140097f
point 1. A singleton class shouldn’t allow other classes to instantiate it

File: src/main/java/mate/academy/jpahw/models/devices/Device.java [hunk 10]
```diff
@@ -0,0 +1,76 @@
...
+    @Id
+    @GeneratedValue(strategy = GenerationType.TABLE)
+    private Long id;
+    @Column
+    private String name;
+    @Column
+    private String model;
+    @Column
+    private String brand;
+
```
Comment (boroda4436) [inline src/main/java/mate/academy/jpahw/models/devices/Device.java]: where is List<Accessory> accessories?

File: src/main/java/mate/academy/jpahw/services/DeviseService.java [hunk 11]
```diff
@@ -0,0 +1,22 @@
+package mate.academy.jpahw.services;
+
+import mate.academy.jpahw.models.acsessory.Acsessory;
+import mate.academy.jpahw.models.devices.Device;
+import mate.academy.jpahw.models.patients.Patient;
+import mate.academy.jpahw.models.tests.Test;
+
+import java.util.List;
+
+public interface DeviseService {
```
Comment (boroda4436) [inline src/main/java/mate/academy/jpahw/services/DeviseService.java]: this is not required, but in furure you may have CrudService, as well as we have CrudDao

File: src/main/java/mate/academy/jpahw/services/FactoryService.java [hunk 12]
```diff
@@ -0,0 +1,37 @@
...
+import mate.academy.jpahw.dao.impl.PatientDaoImpl;
+import mate.academy.jpahw.dao.impl.TestDaoImpl;
+import mate.academy.jpahw.models.acsessory.Acsessory;
+import mate.academy.jpahw.models.devices.Device;
+import mate.academy.jpahw.models.patients.Patient;
+import mate.academy.jpahw.models.tests.Test;
+
+import javax.persistence.EntityManager;
+
+public class FactoryService {
```
Comment (boroda4436) [inline src/main/java/mate/academy/jpahw/services/FactoryService.java]: +1

**Pull request merged**